state-sync/storage-service/server/src/storage.rs

```diff
@@ -107,6 +121,107 @@ impl StorageReader {
+            TransactionListWithProofV2::new(transaction_list_with_auxiliary_info);
+
+        // Return the transaction data response
+        Ok(TransactionDataWithProofResponse {
+            transaction_data_response_type: TransactionDataResponseType::TransactionData,
+            transaction_list_with_proof: Some(transaction_list_with_proof),
+            transaction_output_list_with_proof: None,
+        })
+    }
+
+    /// Constructs the transaction output list with proof v2 (which includes
+    /// auxiliary information for each item in the given list with proof v1).
+    fn construct_output_list_with_proof_v2(
+        &self,
+        transaction_output_list_with_proof: TransactionOutputListWithProof,
+    ) -> Result<TransactionDataWithProofResponse, Error> {
+        // Verify that the first transaction output version exists
+        let first_transaction_output_version = transaction_output_list_with_proof
+            .first_transaction_output_version
+            .ok_or_else(|| {
+                Error::UnexpectedErrorEncountered(
+                    "First transaction output version is missing in the response!".into(),
+                )
+            })?;
+
+        // Get the persisted auxiliary infos
+        let num_infos_to_fetch = transaction_output_list_with_proof
+            .transactions_and_outputs
+            .len();
+        let persisted_auxiliary_infos = self.fetch_persisted_auxiliary_infos(
+            first_transaction_output_version,
+            num_infos_to_fetch,
+        )?;
+
+        // Create the transaction output list with proof v2
+        let transaction_output_list_with_auxiliary_info = TransactionOutputListWithAuxiliaryInfos {
+            transaction_output_list_with_proof,
+            persisted_auxiliary_infos,
+        };
+        let transaction_output_list_with_proof_v2 =
+            TransactionOutputListWithProofV2::new(transaction_output_list_with_auxiliary_info);
+
+        // Return the transaction data response
+        Ok(TransactionDataWithProofResponse {
+            transaction_data_response_type: TransactionDataResponseType::TransactionOutputData,
+            transaction_list_with_proof: None,
+            transaction_output_list_with_proof: Some(transaction_output_list_with_proof_v2),
+        })
+    }
+
+    /// Fetches the persisted auxiliary infos starting at the specified
+    /// version. Note: it is possible for some auxiliary infos to be
+    /// missing, in which case None is returned for each missing version.
+    fn fetch_persisted_auxiliary_infos(
+        &self,
+        first_version: Version,
+        num_infos_to_fetch: usize,
+    ) -> aptos_storage_service_types::Result<Vec<PersistedAuxiliaryInfo>, Error> {
+        // Get an iterator for the persisted auxiliary infos
+        let persisted_auxiliary_info_iter = self
+            .storage
+            .get_persisted_auxiliary_info_iterator(first_version, num_infos_to_fetch)
```

Comment:
question: is it guaranteed the first version is not beyond the ledger version in db?

Reply:
Yeah, there's two ways this is checked:
1. When we receive the request, we ensure we can service it, e.g., [here](https://github.com/aptos-labs/aptos-core/blob/0e7ef13853be9f70a23dffd32c1a8b1a049020a8/state-sync/storage-service/server/src/moderator.rs#L154).
2. We only read the aux infos after we've fetched the transaction/ouput list with proof from storage, so we've verified that we have all the transactions from the first to the last version.

Is this enough, or do you want another check here (just in case)? 😄